include/deal.II/distributed/fullydistributed_tria.h

```diff
@@ -0,0 +1,237 @@
+#  include <p8est_connectivity.h>
+#  include <p8est_ghost.h>
+#endif
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace internal
+{
+  namespace DoFHandlerImplementation
+  {
+    namespace Policy
+    {
+      template <typename>
+      class ParallelDistributed;
+    } // namespace Policy
+  }   // namespace DoFHandlerImplementation
+} // namespace internal
+
+namespace FETools
+{
+  namespace internal
+  {
+    template <int, int, class>
+    class ExtrapolateImplementation;
+  }
+} // namespace FETools
+
+// forward declaration of the data type for periodic face pairs
+namespace GridTools
+{
+  template <typename CellIterator>
+  struct PeriodicFacePair;
+}
+
+namespace parallel
+{
+  namespace fullydistributed
+  {
+    struct CellInfo
+    {
+      CellInfo() = default;
+
+      CellInfo(CellId::binary_type index,
+               unsigned int        subdomain_id,
+               unsigned int        level_subdomain_id)
+        : index(index)
+        , subdomain_id(subdomain_id)
+        , level_subdomain_id(level_subdomain_id){};
+
+      CellId::binary_type index;
+      unsigned int        subdomain_id;
+      unsigned int        level_subdomain_id;
+    };
+
+
+
+    template <int dim, int spacedim>
+    struct ConstructionData
+    {
+      // information describing the local part of the coarse grid
+      std::vector<CellData<dim>>      cells;
+      std::vector<Point<spacedim>>    vertices;
+      std::vector<types::boundary_id> boundary_ids;
+
+      // information
+      std::map<int, int> coarse_lid_to_gid;
+
+      // information describing how to constuct the levels
+      std::vector<std::vector<CellInfo>> parts;
+    };
+
+
+
+    template <int dim, int spacedim = dim>
+    class Triangulation
+      : public parallel::DistributedTriangulationBase<dim, spacedim>
+    {
+    public:
+      typedef typename dealii::Triangulation<dim, spacedim>::cell_iterator
+        cell_iterator;
+
+      typedef
+        typename dealii::Triangulation<dim, spacedim>::active_cell_iterator
+          active_cell_iterator;
+
+      typedef
+        typename dealii::Triangulation<dim, spacedim>::CellStatus CellStatus;
+
+      void
+      reinit(ConstructionData<dim, spacedim> &construction_data);
+
+      virtual void
+      create_triangulation(const std::vector<Point<spacedim>> &vertices,
+                           const std::vector<CellData<dim>> &  cells,
+                           const SubCellData &subcelldata) override;
+
+
+      enum Settings
+      {
+        /**
+         * Default settings, other options are disabled.
+         */
+        default_setting = 0x0,
+        /**
+         * This flags needs to be set to use the geometric multigrid
+         * functionality. This option requires additional computation and
+         * communication. Note: geometric multigrid is still a work in
+         * progress.
+         */
+        construct_multigrid_hierarchy = 0x1
+      };
+
+
+      Triangulation(MPI_Comm mpi_communicator,
```

Comment:
please mark this constructor explicit

Reply:
Done. I guess, I can also create a PR to make the constructor of `p:d:t` explicit?

Reply:
Yes, that's a good idea.